Well typed MongoDB API

ghik opened the pull request:
This PR implements a Scala-idiomatic layer over the Reactive Streams driver for MongoDB.

* `GenCodec` is used for serialization, similarly to already existing `GenCodecCollection` variants.
* Raw `Publisher`s returned by the Reactive Streams driver are translated into Monix `Observable`s and `Task`s.
  * this gives us a very rich API with transformations like `map`, `filter`, etc.
  * it is straightforward to convert between `Observable`s, `Task`s and `Future`s, e.g. we can turn an `Observable[T]` into a `Task[List[T]]` which can then be turned into a `Future[List[T]]`
* Well typed representations of various BSON documents used in database commands: queries, projections, updates, sort orders, indices, etc.
  * Instead of using raw `Bson`, we have `MongoDocumentFilter[E]`, `MongoProjection[E, T]`, `MongoUpdate[E]`, `MongoDocumentOrder[E]`, `MongoIndex[E]` (`E` denotes type of the entity)
* References to document fields are well typed as `MongoPropertyRef[E, T]` (similarly to already existing `BsonRef[E, T]`)
  * just like with `BsonRef`, you can create these references without using raw strings, e.g. `MyEntity.ref(_.myField)`
  * property references directly expose rich API for creating filters, projections, updates, etc.
* Better support for sealed hierarchies
  * for example, it's easy to create a query or projection that limits results to selected case class or intermediate subtrait in a sealed hierarchy

Before reviewing the actual code, one might want to look at the [tutorial](https://github.com/AVSystem/scala-commons/blob/mongo-api/docs/TypedMongo.md).
Good entry point for review is [TypedMongoCollection](https://github.com/AVSystem/scala-commons/pull/317/files#diff-16757a8bf8bf944af4b9d4676b88f0f340b4fc40492a9d5e55c6016a47e1e87b)

ddworak commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/MongoDocumentFilter.scala:
```diff
@@ -0,0 +1,137 @@
+object MongoDocumentFilter {
+  private[this] val reusableEmpty = Empty()
+
+  def empty[E]: MongoDocumentFilter[E] = reusableEmpty.asInstanceOf[MongoDocumentFilter[E]]
```
A filter that's never satisfied is often useful as well (e.g. fold + or). I think `Filters.exists("_id", false)` is always correctly recognized as useless by Mongo when executing such queries, doesn't break index use etc.

ghik replied:
`Filters.exists("_id", false)` will not work for embedded documents.

Due to various random limitations of MongoDB queries I decided to write API only for things which are straightforward to express. Later we may think about emulating some lacking elements like "always false" filter.

ddworak replied:
That's fine by me, I'd just love to have a way to construct a generic method returning such `MongoDocumentFilter[T]` which is probably impossible on the library client side (without taking a ref / companion arg)

ddworak commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/MongoIndex.scala:
```diff
@@ -0,0 +1,50 @@
+import org.bson.{BsonDocument, BsonValue}
+
+
+case class MongoIndex[E](fields: Vector[(MongoPropertyRef[E, _], MongoIndexType)]) {
```
What about index properties? https://docs.mongodb.com/manual/core/index-properties/

ghik replied:
These are passed separately in `IndexOptions`.

Yes, we could rewrite `MongoIndex` to better express various index types along with their options but I decided it would be too much work for now.

ghik commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/Utils.scala:
```diff
@@ -0,0 +1,14 @@
+object Utils {
+  implicit class PublisherOps[T](private val publisher: Publisher[T]) extends AnyVal {
+    def asMonix: Observable[T] = Observable.fromReactivePublisher(publisher)
+    def asMonixTask: Task[T] = Observable.fromReactivePublisher(publisher).firstL
```
Unused, will remove.

ddworak commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/TypedMongoCollection.scala:
```diff
@@ -0,0 +1,292 @@
+import org.bson.{BsonDocument, BsonValue}
+import org.reactivestreams.Publisher
+
+final class TypedMongoCollection[E <: BaseMongoEntity : MongoAdtFormat](
```
You wouldn't expect me to say this, but I'd consider making this non-final. There are some cases where extending this with a bit of logic could serve as a simple yet complete persistence API, without further layers of composition.

ghik replied:
I understand but if you want me to maintain binary compat in a proper way, that might become annoying.

ddworak replied:
That's up to you to decide then, I'm not expecting too much in terms of compatibility when I'm upgrading commons :D

ddworak commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/TypedMongoCollection.scala:
```diff
@@ -0,0 +1,292 @@
+    setupOptions: BulkWriteOptions => BulkWriteOptions = identity
+  ): Task[BulkWriteResult] = {
+    val requests = writes.iterator.map(_.toWriteModel).to[JList]
+    single(nativeCollection.bulkWrite(requests, setupOptions(new BulkWriteOptions)))
```
You could consider skipping the call if `writes` is empty - the native driver notoriously throws exceptions in such case which is usually not what the user expects.

ghik replied:
How do I create `BulkWriteResult` to represent that?

ddworak replied:
You probably don't. Since you're wrapping the API, you can always introduce your own result type.

ghik replied:
But I want to expose `BulkWriteResult` since it contains useful information. Do you want me to wrap it into an `Opt` or something?

ddworak replied:
I wouldn't mind that, basically yes.

ghik replied:
I don't like this because it would force pattern matching or some other empty `Opt` handling in situations where the user is sure that `writes` are non-empty. Also, this goes against the "wrap with better API but don't change semantics" approach of `TypedMongoCollection`.

ddworak commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/TypedMongoCollection.scala:
```diff
@@ -0,0 +1,292 @@
+
+  def bulkWrite(
+    writes: Seq[MongoWrite[E]],
+    setupOptions: BulkWriteOptions => BulkWriteOptions = identity
```
How about an `ordered(false)` default? It's huge for performance, especially when sharding is involved.

ddworak commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/TypedMongoCollection.scala:
```diff
@@ -0,0 +1,292 @@
+
+  def createIndex(
+    index: MongoIndex[E],
+    setupOptions: IndexOptions => IndexOptions = identity
```
If we can afford to be opinionated, please add `background(true)`

ghik replied:
Same remark as before. This would break the assumption that `TypedMongoCollection` is just a better-typed wrapper over the native collection, without changing semantics.

ddworak commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/TypedMongoCollection.scala:
```diff
@@ -0,0 +1,292 @@
+    }.to[JList]
+    single(nativeCollection.createIndexes(indexModels, setupOptions(new CreateIndexOptions)))
+  }
+}
```
Would feel so good to have a nicer `.aggregate` API (or at least validate it fits with your model in the future). It's easily the ugliest, least safe Mongo code we have in our codebase.

ghik replied:
Yeah, I haven't looked into it yet but I'm not sure if we can express pipelines in any better-typed way.

ddworak commented on docs/TypedMongo.md:
```diff
@@ -0,0 +1,446 @@
+* a sealed trait/class with `@flatten` annotation
+
+However, an embedded document is not a toplevel entity and does not need an ID. Therefore, it doesn't need to
+extend `MongoEntity` and its companion must extend `MongoDataCompanion` rather than `MongoEntityCompanion`.
```
Sometimes you really want to use an external type (even though it may be considered a bad practice). If it's possible to do that with this driver, maybe we could be less strict with the `must` and mention how to do this neatly.

ghik replied:
You want to use an external type as an _entity_ type or just _field_ type? As explained in this doc, you can use any type that has `GenCodec` as a field type (opaque).

ddworak replied:
In my case it's field type (I also consider e.g. cross-compiled classes as external in this case). Just wanted to clarify that `must extend` is not really true.

ghik replied:
`must extend` is true for an _embedded document_ whose internal structure is transparent. This doc section is about these.

ddworak commented on docs/TypedMongo.md:
```diff
@@ -0,0 +1,446 @@
+
+`MongoPropertyRef` is a fundamental type that serves as an entry point for creating queries, projections, updates,
+sort orders, indices, etc.
+
```
Would be nice to mention `MongoRef` and `MongoDataRef`, even if just to signal differences with this type.

ghik replied:
These differences are explained in scaladoc. I don't want the tutorial to describe too many rarely used details like `MongoRef` and `MongoDataRef`. These don't really appear in everyday-use API.

ddworak commented on docs/TypedMongo.md:
```diff
@@ -0,0 +1,446 @@
+A `MongoIndex[E]` represents a MongoDB index document for an entity of type `E`. Examples:
+
+* Single field index, e.g. `MyEntity.ref(_.int).ascendingIndex`
+* Multi-field index, e.g.
```
They're usually called compound in the context of MongoDB

ddworak commented on docs/TypedMongo.md:
```diff
@@ -0,0 +1,446 @@
+For more examples of database operations with `TypedMongoCollection`, see
+[tests](https://github.com/AVSystem/scala-commons/blob/mongo-api/commons-mongo/src/test/scala/com/avsystem/commons/mongo/typed/TypedMongoCollectionTest.scala).
+
+`TypedMongoCollection` exposes mostly the same operations as Reactive Streams `MongoCollection` but typed differently, 
```
This should be mentioned much earlier in these docs, after all the library doesn't make much sense without it :P

ddworak commented on docs/TypedMongo.md:
```diff
@@ -0,0 +1,446 @@
+Then we have no choice but to wait for the database on the current JVM thread.
+
+In order to do this, use utilities provided by `com.avsystem.commons.concurrent.BlockingUtils` class. Your project should
+provide an implementation of this class (often a globally accessible object) where an appropriate Monix `Scheduler`s
```
Can we **not** recommend using an `object` for this? Statically defined thread pools usually mean 0 runtime configuration for this.

ghik replied:
Well, I have actually used a globally accessible object for this...
I'm currently under impression that thread pools are one of the rare things that can be kept in globals because in order to  make sure that they are reused throughout your _entire process_. The whole point of using asynchronous API is to reduce the number of threads to minimum.

If we need compatibility then... JVM properties? (I would not think I'd ever say that).

ddworak replied:
I know it's tempting, but injecting some object which provides blocking capabilities **and** is a Scheduler / EC / TP is not that much of a hassle and makes (us) developers reconsider the way we run async tasks. Globally accessible object can also be messy in tests, it's very hard to control or mock it then.

ghik replied:
I've removed the "globally accessible object" suggestion.

ddworak commented on docs/TypedMongo.md:
```diff
@@ -0,0 +1,446 @@
+  // some Scheduler usually reused throughout your application
+  def scheduler: Scheduler = ??? 
+  // some Scheduler usually reused throughout your application (unbounded, for blocking code)
+  def ioScheduler: Scheduler = ??? 
```
Some general recommendations are important here.

ghik commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/DataTypeDsl.scala:
```diff
@@ -0,0 +1,207 @@
+import com.avsystem.commons.annotation.{explicitGenerics, macroPrivate}
+import com.avsystem.commons.macros.serialization.MongoMacros
+
+trait DataRefDsl[E, T] {
```
I will rename `MongoDataRef` to `MongoToplevelRef`

ghik commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/MongoFilter.scala:
```diff
@@ -0,0 +1,152 @@
+  def on[E](prefix: MongoPropertyRef[E, T]): MongoDocumentFilter[E] =
+    PropertyValueFilter(prefix, this)
+
+  final def toBson: BsonDocument = this match {
```
For sealed hierarchies, I usually prefer pattern matching to abstract methods with implementations (more "functional" approach as opposed to "object oriented" one).

ghik commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/MongoFormat.scala:
```diff
@@ -0,0 +1,287 @@
+    valueFormat: MongoFormat[V]
+  ) extends MongoFormat[M[K, V]]
+
+  final case class Optional[O, T](
```
I agree but these types should not be used directly. If used directly, my intention was to always use them with prefix, i.e. `MongoFormat.Optional`, etc.

ddworak replied:
If so, you could expose utility methods and encapsulate the types. It's not a huge issue, but there's quite a bit of "internal public" API in this PR

ghik replied:
Yeah, but in order to get rid of this, I would have to add a lot of package-private modifiers. I would also prefer to leave a possibility to use this internal-ish API in corner cases.

ddworak commented on docs/TypedMongo.md:
````diff
@@ -0,0 +1,447 @@
+  ```scala
+  import MongoIndexType._
+  val index: MongoIndex[MyEntity] =
+  MongoIndex(MyEntity.ref(_.str) -> Hashed, MyEntity.ref(_.int) -> Descending)
````
This would actually be an error (hashed indexes cannot be part of a compound index). Maybe the API is too liberal?

ghik replied:
Yeah I should rethink `MongoIndex`

ghik replied:
Well that's actually not true anymore: https://docs.mongodb.com/manual/core/index-hashed/#create-a-compound-hashed-index

ghik replied:
I know that this API still allows creation of many invalid indexes but I think there's no point in trying to cover this with very precise static API - it will likely be complicated and become incompatible as MongoDB is evolving.

ddworak replied:
Whoa that's fresh. I'm okay with this being a bit liberal then, but I'm still not the biggest fan of this API in the sense that single indexes are created via extension methods on refs and compound indexes have a completely different feel (making them harder to discover I guess?). Maybe that's a good thing though given how different such indexes are.

ghik commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/MongoQueryOperator.scala:
```diff
@@ -0,0 +1,78 @@
+  final case class Exists[T](exists: Boolean) extends MongoQueryOperator[T]
+  final case class Type[T](bsonType: BsonType) extends MongoQueryOperator[T]
+  final case class Regex[T](pattern: String) extends MongoQueryOperator[T] //TODO: options
+  final case class Mod[T](divisor: Long, remainder: Long) extends MongoQueryOperator[T]
```
I tried using `com.mongodb.client.model.Filters` and it [failed](https://stackoverflow.com/questions/64557424/how-do-i-express-complex-elemmatch-mongodb-query-using-java-driver).

I believe that the compatibility layer you're talking about is the already available `Raw` operator.

ghik commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/MongoRef.scala:
```diff
@@ -0,0 +1,329 @@
+  *
+  * @tparam E the data type
+  */
+sealed trait MongoRef[E, T] extends MongoProjection[E, T] with DataRefDsl[E, T] { self =>
```
I agree but also I don't like `MyRef.filter.is(value)` as opposed to `MyRef.is(value)`.
Is this any worse that the current situation with `BsonRef`?

ghik commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/MongoRef.scala:
```diff
@@ -0,0 +1,329 @@
+  def on[E0](ref: MongoRef[E0, E]): MongoProjection[E0, T] = compose(ref)
+}
+
+sealed trait MongoDataRef[E, T <: E] extends MongoRef[E, T] {
```
Sure, I'll add some but this type is not really used in the public API.

ghik commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/MongoRef.scala:
```diff
@@ -0,0 +1,329 @@
+  def format: MongoFormat[T]
+  def projectionRefs: Set[MongoRef[E, _]] = Set(this)
+  def showRecordId: Boolean = false
+
```
This gets a bit complicated because `MongoRef` is technically a _partial_ function now. Also, for consistency I would probably have to add this on `MongoProjection` level and maybe `MongoFilter` should also have a memory-level filter, etc...

ddworak commented on commons-core/jvm/src/main/scala/com/avsystem/commons/concurrent/ObservableBlockingIterator.scala:
```diff
@@ -0,0 +1,98 @@
+  * This essentially turns an `Observable` into an `Iterator`
+  * (which requires blocking so use this only as a last resort).
+  */
+class ObservableBlockingIterator[T](
```
I've seen an alternative implementation in RxJava `https://github.com/ReactiveX/RxJava/blob/3.x/src/main/java/io/reactivex/rxjava3/internal/operators/observable/BlockingObservableIterable.java`. You might want to take a look or borrow some tests for better coverage :eyes:

ghik replied:
Quite a different implementation though because this one isn't based on Monix

ghik commented on commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/MongoFormat.scala:
```diff
@@ -0,0 +1,287 @@
+
+/**
+  * Typeclass that captures internal structure of a type that can be saved to MongoDB
+  * (directly as a toplevel entity or indirectly as an embedded value).
```
I'll add a section to the tutorial about this.

ghik commented:
@ddworak I think I've addressed all your concerns, either with code or with a reply. Some of these replies are in the spirit of "I'd prefer avoid changing this but I will if you think it's important".

ddworak commented on docs/TypedMongo.md:
```diff
@@ -444,3 +448,23 @@ Using this utility, you can:
+In comparison to the old API, the new one provides:
+
+* `TypedMongoCollection` wrapper over Reactive Streams collection - as opposed to native `MongoCollection`s
+  * other drivers are unsupported because they are deprecated, synchronous or redundant
```
> deprecated, synchronous or redundant

 I love it.

ddworak reviewed (approved):
LGTM, this is a significant improvement over previous approaches and fits nicely with our ecosystem. The leftover 3 comments are up to you to decide :shipit:

ghik commented:
Thanks a lot! I will merge this in current form in order to push things further in downstream projects.
